pkg/datastore/database.go

```diff
@@ -0,0 +1,771 @@
+		return nil, err
+	}
+
+	sqlDB, err := db.DB()
+	if err != nil {
+		return nil, err
+	}
+
+	// Ensure DB connection works
+	if err = sqlDB.Ping(); err != nil {
+		return nil, err
+	}
+
+	return db, nil
+}
+
+func TypeName(x interface{}) (typeName string) {
+	t := reflect.TypeOf(x)
+	for t.Kind() == reflect.Ptr {
+		t = t.Elem()
+		typeName += "*"
+	}
+	typeName += t.Name()
+	return
+}
+
+func IsPointerToStruct(x interface{}) (isPtrType bool) {
+	t := reflect.TypeOf(x)
+	isPtrType = (t.Kind() == reflect.Ptr && t.Elem().Kind() == reflect.Struct)
+	return
+}
+
+// Validates tenancy information and returns a transaction with right dbRole.
+func (db *relationalDb) GetDBTransaction(ctx context.Context, tableName string, record Record) (tx *gorm.DB, err error) {
+	if !IsPointerToStruct(record) {
+		return nil, ErrNotPtrToStruct.WithValue(TYPE, TypeName(record))
+	}
+
+	return db.getDBTransaction(ctx, tableName, record)
+}
+
+func (db *relationalDb) getDBTransaction(ctx context.Context, tableName string, record Record) (tx *gorm.DB, err error) {
+	err, orgId, dbRole, isDbRoleTenantScoped := db.getTenantInfoFromCtx(ctx, tableName)
+	if err != nil {
+		return nil, err
+	}
+
+	// If the DB role is tenant-specific (TENANT_READER or TENANT_WRITER) and the table is multi-tenant,
+	// make sure that the record being inserted/modified/updated/deleted/queried belongs to the user's org.
+	// If operation is SELECT but no specific tenant's data is being queried (e.g., FindAll() was called), allow the operation to proceed.
+	if dbRole.IsDbRoleTenantScoped() && IsMultitenant(record, tableName) {
+		orgIdCol, _ := GetOrgId(record)
+		if orgIdCol != "" && orgIdCol != orgId {
+			err = ErrOperationNotAllowed.WithValue("tenant", orgId).WithValue("orgIdCol", orgIdCol)
+			db.logger.Error(err)
+			return nil, err
+		}
+	}
+
+	tx = db.gormDBMap[dbRole].Begin()
+	if isDbRoleTenantScoped {
+		// Set org. ID
+		stmt := getSetConfigStmt(DB_CONFIG_ORG_ID, orgId)
+		if tx = tx.Exec(stmt); tx.Error != nil {
+			err = ErrExecutingSqlStmt.Wrap(tx.Error).WithValue(SQL_STMT, stmt)
+			db.logger.Errorln(err)
+			return nil, err
+		}
+	}
+	tx = tx.Table(tableName)
+	if tx.Error != nil {
+		err = ErrStartingTx.Wrap(tx.Error).WithMap(map[ErrorContextKey]string{
+			"db_role":    string(dbRole),
+			"table_name": tableName,
+			"authorizer": TypeName(db.authorizer),
+		})
+		db.logger.Error(err)
+		return nil, err
+	}
+
+	tx = tx.Table(tableName)
+	return tx, nil
+}
+
+// Resets DB connection pools.
+func (db *relationalDb) Reset() {
+	db.gormDBMap = make(map[dbrole.DbRole]*gorm.DB)
+	db.authorizer = authorizer.MetadataBasedAuthorizer{}
+	if err := db.Initialize(); err != nil {
+		panic(err)
+	}
+}
+
+// Finds a single record that has the same primary key as record.
+// record argument must be a pointer to a struct and will be modified in-place.
+// Returns ErrRecordNotFound if a record could not be found.
+func (db *relationalDb) Find(ctx context.Context, record Record) error {
```

Comment:
Maybe we could rename the `record` argument to `filter`?

Reply:
leaving as record as it is the output too.